Userland/Libraries/LibJS/Runtime/PromiseCapability.cpp

```diff
@@ -23,32 +23,30 @@ ThrowCompletionOr<PromiseCapability> new_promise_capability(VM& vm, Value constr
     // 2. NOTE: C is assumed to be a constructor function that supports the parameter conventions of the Promise constructor (see 27.2.3.1).
 
     // 3. Let promiseCapability be the PromiseCapability Record { [[Promise]]: undefined, [[Resolve]]: undefined, [[Reject]]: undefined }.
-    // FIXME: This should not be stack-allocated, the executor function below can be captured and outlive it!
-    //        See https://discord.com/channels/830522505605283862/886211697843531866/900081190621569154 for some discussion.
     struct {
-        Value resolve { js_undefined() };
-        Value reject { js_undefined() };
+        Handle<Value> resolve;
+        Handle<Value> reject;
     } promise_capability_functions;
 
     // 4. Let executorClosure be a new Abstract Closure with parameters (resolve, reject) that captures promiseCapability and performs the following steps when called:
-    auto executor_closure = [&promise_capability_functions](auto& vm) -> ThrowCompletionOr<Value> {
+    auto executor_closure = [promise_capability_functions](auto& vm) mutable -> ThrowCompletionOr<Value> {
```

Comment:
Don't we still need to capture `promise_capability_functions` by reference?
Now we are setting things on the copy that `executor_closure` has.

But if this works I would love to know what I'm missing here.

Reply:
We're no longer capturing by reference, we're capturing by value :)

This means each member (a `Handle`) gets copied, internally that just increases the refcount of each `HandleImpl`, thus keeping them alive as long as this closure exists.

Reply:
...or so I thought 😂
Since Handle<Value> is a specialization, that plan doesn't work. Back to the drawing board...

Reply:
@davidot I implemented it in a slightly different way now, PTAL!